src/non_RL_algos/shortestPath.py

```diff
@@ -0,0 +1,250 @@
+                    {
+                        'SFC id' : dict
+                        {
+                            'SF id' : dict
+                            {
+                                'node id' : float (Inclusive of zero values)
+                            }
+                        }
+                    }
+                }
+        '''
+
+    Parameters:
+        network: A NetworkX object
+        nodes_list: all the nodes in the network
+        sf_list: all the sf's in the network
+        sfc_list: all the SFCs in the network, right now assuming to be just 1
+        ingress_nodes: all the ingress nodes in the network
+        nodes_cap: Capacity of each node in the network
+
+    Returns:
+        - a placement Dictionary with:
+              key = nodes of the network
+              value = list of all the SFs in the network
+        - schedule of the form shown above
+    """
+    placement = defaultdict(list)
+    schedule = defaultdict(lambda: defaultdict(lambda: defaultdict(lambda: defaultdict(float))))
+    # Initializing the schedule for all nodes, for all SFs to 0
+    for src in nodes_list:
+        for sfc in sfc_list:
+            for sf in sf_list:
+                for dstn in nodes_list:
+                    schedule[src][sfc][sf][dstn] = 0
+    # Finding the closest neighbours to each node in the network. For each node of the network we maintain a list of
+    # neighbours sorted in increasing order of distance to it.
+    all_pair_shortest_paths = network.graph['shortest_paths']
+    closest_neighbours = defaultdict(list)
+    for source in nodes_list:
+        neighbours = defaultdict(int)
+        for dest in nodes_list:
+            if source != dest:
+                delay = all_pair_shortest_paths[(source, dest)][1]
+                neighbours[dest] = delay
+        sorted_neighbours = [k for k, v in sorted(neighbours.items(), key=lambda item: item[1])]
+        closest_neighbours[source] = sorted_neighbours
+
+    # - For each Ingress node of the network we start by placing the first VNF of the SFC on it and then place the
+    #  2nd VNF of the SFC on the closest neighbour of the Ingress, then the 3rd VNF on the closest neighbour of the node
+    #  where we placed the 2nd VNF and so on.
+    # - The closest neighbour is chosen based on the following criteria:
+    #   - while some nodes in the network has 0 VNFs , the closest neighbour cannot be an Ingress node
+    #   - The closest neighbour must have some capacity
+    #   - while some of the nodes in the network have 0 VNFs it chooses the closest neighbour that has 0 VNFs,
+    #     If some nodes in the network has just 1 VNF, it returns the closest neighbour with just 1 VNF and so on
+    for ingress in ingress_nodes:
+        node = ingress
+        # We choose a list with just one element because a list is mutable in python and we want 'next_neighbour'
+        # function to change the value of this variable
+        num_vnfs_filled = [0]
```

Comment:
Why not just return the updated num of vnfs at the end of the function call as 2nd return value? I think that would be clearer

Reply:
I thought about doing that but we are calling the `next_neighbour` function many times and most of those time the `num_vnfs_filled` variable is not even changed. So I think it would just be useless to do that.